### PR TITLE
DO NOT MERGE - discussion of test framework

### PR DESCRIPTION
**Pull request by madeleineth**

### CODE REVIEW

File: tests/util/test_math.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from scipy.stats import norm
+from numpy import log, sqrt, exp, inf
+
+from util import test_statistic_threshold as tst
+
+
+def kullback_leibler(post_sample, post_pdf, comp_pdf):
+    """Estimate KL-divergence of sample (a collection of values) w.r.t. known pdf,
+    comp_pdf, which returns the density when passed a sample. Return value is
+    (estimated_kl, variance) post_sample is an approximate sample from the
```
Comment (madeleineth) [inline tests/util/test_math.py]: Can you clarify what variance this is? Estimated sampling variance of the KL divergence?

File: tests/util/test_math.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    distribution approximately represented by post_pdf
+
+    """
+    klsamples = [log(post_pdf(x)) - log(comp_pdf(x)) for x in post_sample]
+    klsamples = scipy.array(klsamples)
+    # In general the mean and variance of log(P(x)) is not known to be finite,
+    # but it will be for any distribution crosscat generates at the moment,
+    # because they all have finite entropy.  Hence CLT applies
+    return klsamples.mean(), klsamples.var() / sqrt(len(klsamples))
```
Comment (madeleineth) [inline tests/util/test_math.py]: This doesn't look numerically reliable. Consider computing entirely in the log domain.

File: tests/util/test_statistic_threshold.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    than "x" is less than "threshold". "x" is the ML estimate of the
+    threshold**(1/ns) quantile"""
+
+    # Type checking
+    if not all(isinstance(d, numbers.Number) for d in observed):
+        raise ValueError, 'observed is not a list of numeric values'
+    if (threshold > 1) or (threshold < 0):
+        raise ValueError, 'threshold is not a probability'
+    if (round(ns) != ns) or (ns < 1):
+        raise ValueError, 'ns is not a natural number'
```
Comment (madeleineth) [inline tests/util/test_statistic_threshold.py]: FYI: This syntax is deprecated and won't work in Python 3. Instead:

```
raise ValueError('ns is not a natural number')
```

**Comment by alxempirical** [comment]
Thanks for the feedback, @madeleineth !

**Pull request closed without merging**